### PR TITLE
fix(udf): check the data type returned from UDF server

### LINKED ISSUE #10828: UDF panics if the function signature was modified discreetly.
### Describe the bug

There's only check in frontend, but not expr, so we can restart the function server and change function signature.

```python
@udf(input_types=["VARCHAR"], result_type="VARCHAR")
def foo(x) -> int:
    return "42"
```
Then create function
```sql
dev=> create function bar(varchar) returns varchar as 'foo' using link 'http://localhost:8815';
CREATE_FUNCTION
dev=> select bar('1');
 bar 
-----
 42
(1 row)

```

Then change code and restart server
```python
@udf(input_types=["VARCHAR"], result_type="INT")
def foo(x) -> int:
    return 42
```

Then boom 💥
```sql
dev=> select bar('1');
ERROR:  Panicked when processing: internal error: entered unreachable code
```

### Error message/log

```text
thread 'risingwave-main' panicked at 'internal error: entered unreachable code', src/common/src/types/to_text.rs:107:1
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panicking.rs:593:5
   1: core::panicking::panic_fmt
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/panicking.rs:67:14
   2: core::panicking::panic
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/panicking.rs:117:5
   3: <i32 as risingwave_common::types::to_text::ToText>::write_with_type
             at ./src/common/src/types/to_text.rs:94:30
   4: <risingwave_common::types::ScalarRefImpl as risingwave_common::types::to_text::ToText>::write_with_type
             at ./src/common/src/types/to_text.rs:207:58
   5: risingwave_common::types::to_text::ToText::to_text_with_type
             at ./src/common/src/types/to_text.rs:34:9
   6: risingwave_common::types::ScalarRefImpl::text_format
             at ./src/common/src/types/mod.rs:1055:9
   7: risingwave_frontend::handler::util::pg_value_format
             at ./src/frontend/src/handler/util.rs:126:20
   8: risingwave_frontend::handler::util::to_pg_rows::{{closure}}::{{closure}}
             at ./src/frontend/src/handler/util.rs:165:40
   9: core::iter::adapters::map::map_try_fold::{{closure}}
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/iter/adapters/map.rs:91:28
  10: core::iter::traits::iterator::Iterator::try_fold
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/iter/traits/iterator.rs:2303:21
  11: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::try_fold
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/iter/adapters/map.rs:117:9
  12: <core::iter::adapters::GenericShunt<I,R> as core::iter::traits::iterator::Iterator>::try_fold
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/iter/adapters/mod.rs:195:9
  13: core::iter::traits::iterator::Iterator::try_for_each
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/iter/traits/iterator.rs:2365:9
  14: <core::iter::adapters::GenericShunt<I,R> as core::iter::traits::iterator::Iterator>::next
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/iter/adapters/mod.rs:178:14
  15: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/alloc/src/vec/spec_from_iter_nested.rs:26:32
  16: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/alloc/src/vec/spec_from_iter.rs:33:9
  17: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/alloc/src/vec/mod.rs:2711:9
  18: core::iter::traits::iterator::Iterator::collect
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/iter/traits/iterator.rs:1895:9
  19: <core::result::Result<V,E> as core::iter::traits::collect::FromIterator<core::result::Result<A,E>>>::from_iter::{{closure}}
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/result.rs:1932:51
  20: core::iter::adapters::try_process
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/iter/adapters/mod.rs:164:17
  21: <core::result::Result<V,E> as core::iter::traits::collect::FromIterator<core::result::Result<A,E>>>::from_iter
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/result.rs:1932:9
  22: core::iter::traits::iterator::Iterator::collect
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/iter/traits/iterator.rs:1895:9
  23: itertools::Itertools::try_collect
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/itertools-0.10.5/src/lib.rs:2014:9
  24: risingwave_frontend::handler::util::to_pg_rows::{{closure}}
             at ./src/frontend/src/handler/util.rs:160:23
  25: core::iter::adapters::map::map_try_fold::{{closure}}
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/iter/adapters/map.rs:91:28
  26: core::iter::traits::iterator::Iterator::try_fold
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/iter/traits/iterator.rs:2303:21
  27: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::try_fold
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/iter/adapters/map.rs:117:9
  28: <core::iter::adapters::GenericShunt<I,R> as core::iter::traits::iterator::Iterator>::try_fold
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/iter/adapters/mod.rs:195:9
  29: core::iter::traits::iterator::Iterator::try_for_each
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/iter/traits/iterator.rs:2365:9
  30: <core::iter::adapters::GenericShunt<I,R> as core::iter::traits::iterator::Iterator>::next
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/iter/adapters/mod.rs:178:14
  31: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/alloc/src/vec/spec_from_iter_nested.rs:26:32
  32: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/alloc/src/vec/spec_from_iter.rs:33:9
  33: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/alloc/src/vec/mod.rs:2711:9
  34: core::iter::traits::iterator::Iterator::collect
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/iter/traits/iterator.rs:1895:9
  35: <core::result::Result<V,E> as core::iter::traits::collect::FromIterator<core::result::Result<A,E>>>::from_iter::{{closure}}
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/result.rs:1932:51
  36: core::iter::adapters::try_process
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/iter/adapters/mod.rs:164:17
  37: <core::result::Result<V,E> as core::iter::traits::collect::FromIterator<core::result::Result<A,E>>>::from_iter
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/result.rs:1932:9
  38: core::iter::traits::iterator::Iterator::collect
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/iter/traits/iterator.rs:1895:9
  39: itertools::Itertools::try_collect
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/itertools-0.10.5/src/lib.rs:2014:9
  40: risingwave_frontend::handler::util::to_pg_rows
             at ./src/frontend/src/handler/util.rs:155:5
  41: <risingwave_frontend::handler::util::DataChunkToRowSetAdapter<VS> as futures_core::stream::Stream>::poll_next
             at ./src/frontend/src/handler/util.rs:101:25
  42: futures_util::stream::stream::StreamExt::poll_next_unpin
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.28/src/stream/stream/mod.rs:1632:9
  43: <risingwave_frontend::handler::PgResponseStream as futures_core::stream::Stream>::poll_next
             at ./src/frontend/src/handler/mod.rs:92:52
  44: futures_util::stream::stream::StreamExt::poll_next_unpin
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.28/src/stream/stream/mod.rs:1632:9
  45: <futures_util::stream::stream::next::Next<St> as core::future::future::Future>::poll
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.28/src/stream/stream/next.rs:32:9
  46: pgwire::pg_protocol::PgProtocol<S,SM>::inner_process_query_msg_one_stmt::{{closure}}
             at ./src/utils/pgwire/src/pg_protocol.rs:444:66
  47: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  48: pgwire::pg_protocol::PgProtocol<S,SM>::inner_process_query_msg::{{closure}}
             at ./src/utils/pgwire/src/pg_protocol.rs:401:18
  49: pgwire::pg_protocol::PgProtocol<S,SM>::process_query_msg::{{closure}}
             at ./src/utils/pgwire/src/pg_protocol.rs:365:65
  50: pgwire::pg_protocol::PgProtocol<S,SM>::do_process_inner::{{closure}}
             at ./src/utils/pgwire/src/pg_protocol.rs:244:88
  51: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::future::future::Future>::poll
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/panic/unwind_safe.rs:296:9
  52: <futures_util::future::future::catch_unwind::CatchUnwind<Fut> as core::future::future::Future>::poll::{{closure}}
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.28/src/future/future/catch_unwind.rs:36:42
  53: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/panic/unwind_safe.rs:271:9
  54: std::panicking::try::do_call
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panicking.rs:500:40
  55: ___rust_try
  56: std::panicking::try
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panicking.rs:464:19
  57: std::panic::catch_unwind
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panic.rs:142:14
  58: <futures_util::future::future::catch_unwind::CatchUnwind<Fut> as core::future::future::Future>::poll
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.28/src/future/future/catch_unwind.rs:36:9
  59: <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll::{{closure}}
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/task/task_local.rs:348:35
  60: tokio::task::task_local::LocalKey<T>::scope_inner
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/task/task_local.rs:233:19
  61: <tokio::task::task_local::TaskLocalFuture<T,F> as core::future::future::Future>::poll
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/task/task_local.rs:345:13
  62: pgwire::pg_protocol::PgProtocol<S,SM>::do_process::{{closure}}
             at ./src/utils/pgwire/src/pg_protocol.rs:169:14
  63: pgwire::pg_protocol::PgProtocol<S,SM>::process::{{closure}}
             at ./src/utils/pgwire/src/pg_protocol.rs:163:30
  64: pgwire::pg_server::handle_connection::{{closure}}::{{closure}}
             at ./src/utils/pgwire/src/pg_server.rs:173:45
  65: pgwire::pg_server::handle_connection::{{closure}}
             at ./src/utils/pgwire/src/pg_server.rs:158:1
  66: <F as futures_core::future::TryFuture>::try_poll
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-core-0.3.28/src/future.rs:82:9
  67: <futures_util::future::try_future::into_future::IntoFuture<Fut> as core::future::future::Future>::poll
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.28/src/future/try_future/into_future.rs:34:9
  68: <futures_util::future::future::map::Map<Fut,F> as core::future::future::Future>::poll
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.28/src/future/future/map.rs:55:37
  69: <futures_util::future::future::Inspect<Fut,F> as core::future::future::Future>::poll
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.28/src/lib.rs:91:13
  70: <futures_util::future::try_future::InspectErr<Fut,F> as core::future::future::Future>::poll
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.28/src/lib.rs:91:13
  71: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/future/future.rs:125:9
  72: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
  73: tokio::runtime::task::core::Core<T,S>::poll::{{closure}}
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/core.rs:223:17
  74: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/loom/std/unsafe_cell.rs:14:9
  75: tokio::runtime::task::core::Core<T,S>::poll
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/core.rs:212:13
  76: tokio::runtime::task::harness::poll_future::{{closure}}
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/harness.rs:476:19
  77: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/core/src/panic/unwind_safe.rs:271:9
  78: std::panicking::try::do_call
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panicking.rs:500:40
  79: ___rust_try
  80: std::panicking::try
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panicking.rs:464:19
  81: std::panic::catch_unwind
             at /rustc/f0411ffcebcd7f75ac02ed45feb53ffd07b75398/library/std/src/panic.rs:142:14
  82: tokio::runtime::task::harness::poll_future
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/harness.rs:464:18
  83: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/harness.rs:198:27
  84: tokio::runtime::task::harness::Harness<T,S>::poll
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/harness.rs:152:15
  85: tokio::runtime::task::raw::poll
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/raw.rs:255:5
  86: tokio::runtime::task::raw::RawTask::poll
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/raw.rs:200:18
  87: tokio::runtime::task::LocalNotified<S>::run
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/task/mod.rs:394:9
  88: tokio::runtime::scheduler::multi_thread::worker::Context::run_task::{{closure}}
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/scheduler/multi_thread/worker.rs:464:13
  89: tokio::runtime::coop::with_budget
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/coop.rs:107:5
  90: tokio::runtime::coop::budget
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/coop.rs:73:5
  91: tokio::runtime::scheduler::multi_thread::worker::Context::run_task
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/scheduler/multi_thread/worker.rs:463:9
  92: tokio::runtime::scheduler::multi_thread::worker::Context::run
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/scheduler/multi_thread/worker.rs:426:24
  93: tokio::runtime::scheduler::multi_thread::worker::run::{{closure}}
             at /Users/xxchan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.0/src/runtime/scheduler/multi_thread/worker.rs:406:17

```

### To Reproduce

_No response_

### Expected behavior

_No response_

### How did you deploy RisingWave?

_No response_

### The version of RisingWave

_No response_

### Additional context

#9002

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

fix #10828

## Checklist

- [x] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests (It seems not easy to test it.)
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [x] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/common/src/types/struct_type.rs [hunk 1]
```diff
@@ -117,6 +117,16 @@ impl StructType {
             .chain(std::iter::repeat("").take(self.0.field_types.len() - self.0.field_names.len()))
             .zip_eq_debug(self.0.field_types.iter())
     }
+
+    /// Compares the datatype with another, ignoring nested field names and metadata.
+    pub fn equals_datatype(&self, other: &StructType) -> bool {
+        if self.0.field_types.len() != other.0.field_types.len() {
```
Comment (xxchan) [inline src/common/src/types/struct_type.rs:123]: nit: zip_eq_fast already contains length check?

**Review by xxchan — APPROVED** [review]
1. Should we add some tests?
2. When the error occurs, eval_infallible will call UDF for each row, right? Not sure if it has any problems in this case. I think it’s not a very big deal. 🤔️

File: src/common/src/types/struct_type.rs (hunk 1 above)
Comment (wangrunji0408) [reply src/common/src/types/struct_type.rs:123]: But it will panic when the length doesn't match, right? We can't panic here.

**Comment by wangrunji0408** [comment]
> 1. Should we add some tests?

Yes. But I couldn't think of a way to test this in slt. 🤔️

> 2. When the error occurs, eval_infallible will call UDF for each row, right? Not sure if it has any problems in this case. I think it’s not a very big deal. 🤔️

Yes. I'm also not sure about it. I'll be soon working on fixing #11586. At that time I may put the responsibility of `eval_infallible` on each function. And for UDF it may simply fill all values to null.

**Review by TennyZhuang — APPROVED** [review]

File: src/common/src/types/struct_type.rs (hunk 1 above)
Comment (xxchan) [reply src/common/src/types/struct_type.rs:123]: You are right

**Pull request merged**